_posts/2018-05-07-is-graphql-the-future.markdown

```diff
@@ -0,0 +1,196 @@
+
+# GraphQL anatomy
+
+A GraphQL always request starts with at least _one root API operation_ and some finite number of follow-up operations. Idiomatically, these follow-up operations are queries, meaning that they just return data, without changing the server state in observable ways. Every operation—-root or otherwise—-can specify some finite number of follow-up operations. GraphQL models API operations as **fields**. How a field works in GraphQL depends on its **type**, which falls into one of two basic categories:
+
+* **Scalar** types (`Int`, `Float`, `String`, `Boolean`, and `ID`, as well as application-defined `enum` and `scalar` types) represent the individual pieces of _data actually sent to the client_. Contrary how I think of the term scalar in other contexts, the data can be arbitrarily complex. As far as the GraphQL spec is concerned, scalars are just opaque blobs of data with validation and serialization rules. As an operation, a scalar field is terminal data fetch, with no follow-ups. They are the leaves of the request tree.
+* **Object** types (`type`, `union` and `interface`) are collections of fields. As an operation, an object-typed field is an intermediate operation that serves as the junction point for follow-up operations. But itself, it doesn’t return any data. They are the branches of the request tree.
+
+The entire model for a given API is known as its **schema**. Every schema has a root query object type, whose fields serve as the API’s entry points.
+
+```graphql
+# The root query object type
+type Query {
+  artwork(id: ID): Artwork
+  artist(name: String)
+  # … a whole bunch more root fields
+}
+
+type Artwork {
+  title: String
+  artist: Artist
+}
+
+type Artist {
+  name: String
+}
+```
+
+A GraphQL request begins by mentioning at least one of the fields of the root query object. This represents an initial query. And if that field is an object, _its_ fields any number of follow-up queries. Critically, _any_ field in the request tree can take arguments, allowing a request to be parameterized at all depths.
+
+Take this query, for example:
+
+```graphql
+{
+  artwork(id: "andy-warhol-campbells-soup-i-black-bean") {
+    title
+    artist {
+      name
+    }
+  }
+}
+```
+
+Here, we tell the server to look up an Artwork by its slug, and tell us the title. So far, this is just like REST. But we _also_ tell it to find us the Artist for us. Importantly, object fields _must_ be followed up with further queries, and scalar fields _cannot_ be. With that in mind, it’s easy to see that `artwork` and `artist` are object fields, while `title` and `name` are scalar fields.
```

Comment:
"But we _also_ tell it to find us the Artist for us" -> "But we _also_ tell it to find the `Artist` (type) for us".